### PR TITLE
Zsh completion for docker-compose

### PR DESCRIPTION
**Pull request by sdurrheimer**
Here my work on a zsh completion script with updated completion.md doc.

Signed-off-by: Steve Durrheimer <s.durrheimer@gmail.com>

### CODE REVIEW

**Comment by sdurrheimer** [comment]
Anyone ? :confused:

**Comment by aanand** [comment]
Thanks for this! I'm going to have a look at it soon - I'll be its first customer.

Can any of the bash and zsh completion be factored out to a common set of functions?

**Comment by sdurrheimer** [comment]
Some of them might be, but this will make the installation process more complex forcing people to install 2 files (not in the same directory). This will therefore create a dependency of a file to another.

Maybe @albers  can also give us his opinion.

**Comment by albers** [comment]
Cool, zsh completion finally arrived. Thanks @sdurrheimer!

I'm not familiar with either using or coding zsh completion, but looking at the code I see a lot of potential for shared code. I like the idea to create a shared lib in order to bundle the efforts for both completions. 

But I also share sdurrheimers concerns about the increased complexity of the installation process. Where should the shared lib be placed? People will run into problems with file permissions.
This would be different if docker-compose had native installer packages like docker has.

More important, I'm not in favour of creating a shared lib unless there are tests.
I was unsuccessful in creating tests for the bash completion. Splitting out functionality into a shared lib increases the danger of introducing regressions. This is exceptionally the case when the lib is used by different "languages" / "frameworks" / "envrionments". Personally, I would feel very uncomfortable working on the shared lib due to my lack of zsh skills.

Summing up, I'm against the shared lib.

**Comment by bfirsh** [comment]
Needs a rebase - sorry @sdurrheimer. :(

**Comment by albers** [comment]
@sdurrheimer 
Hi Steve, there's a new command `migrate_to_labels`, see #1432.
It is top-level and has neither options nor args.

**Comment by sdurrheimer** [comment]
@albers Ok, will do.

@bfirsh Rebase done and commit squashed. I will keep it up to date if I see other commits.

File: docs/completion.md [hunk 1]
```diff
@@ -3,23 +3,44 @@ layout: default
...
 Make sure bash completion is installed. If you use a current Linux in a non-minimal installation, bash completion should be available.
 On a Mac, install with `brew install bash-completion`
- 
-Place the completion script in `/etc/bash_completion.d/` (`/usr/local/etc/bash_completion.d/` on a Mac), using e.g. 
 
-     curl -L https://raw.githubusercontent.com/docker/compose/1.2.0/contrib/completion/bash/docker-compose > /etc/bash_completion.d/docker-compose
- 
+Place the completion script in `/etc/bash_completion.d/` (`/usr/local/etc/bash_completion.d/` on a Mac), using e.g.
+
+     curl -L https://raw.githubusercontent.com/docker/compose/$(docker-compose --version | awk '{print $2}')/contrib/completion/bash/docker-compose > /etc/bash_completion.d/docker-compose
```
Comment (albers) [inline docs/completion.md]: Nice idea to compute the URL here. Please adjust the _Release process_ section in [CONTRIBUTING.md](https://github.com/docker/compose/blob/master/CONTRIBUTING.md#release-process). It says the script URL should be updated.

**Comment by sdurrheimer** [comment]
I just added --x-smart-recreate and -v. Thx @albers.

**Comment by albers** [comment]
@aanand @bfirsh Any chance to get zsh completion this into 1.7?

**Comment by aanand** [comment]
LGTM

I've given this a quick try and I like it a lot. Would be nice to get it in for 1.7, but I don't know if it should count as code for the purposes of "code freeze" - which was three days ago.

**Comment by bfirsh** [comment]
Naah it's tooling. The code freeze is for features.

**Comment by bfirsh** [comment]
LGTM

**Comment by aanand** [comment]
zorro

**Comment by albers** [comment]
@bfirsh 

> Naah it's tooling. The code freeze is for features.

This is an important point for me. Although I try to keep up with new features that are relevant for completion, I always end up with finding some in the release notes of the RC.

cc @jfrazelle @tianon - also applies to Docker completion.

**Comment by tianon** [comment]
I know I'm late to the discussion (:innocent:), but I'm definitely +1 to including fixes and improvements to things like completion even after code freeze since it makes for a more polished release overall. :+1:
